x-pack/plugins/security_solution/public/management/hooks/artifacts/test_utils.tsx

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import React from 'react';
+import { renderHook } from '@testing-library/react-hooks';
+import { QueryClient, QueryClientProvider } from 'react-query';
+import { CoreStart, HttpSetup } from 'kibana/public';
+import { CreateExceptionListSchema } from '@kbn/securitysolution-io-ts-list-types';
+import { coreMock } from '../../../../../../../src/core/public/mocks';
+
+export const getFakeListId: () => string = () => 'FAKE_LIST_ID';
+export const getFakeListDefinition: () => CreateExceptionListSchema = () => ({
+  name: 'FAKE_LIST_NAME',
+  namespace_type: 'agnostic',
+  description: 'FAKE_LIST_DESCRIPTION',
+  list_id: getFakeListId(),
+  type: 'endpoint',
+});
+
+export const getFakeHttpService = () => {
+  const fakeCoreStart = coreMock.createStart({ basePath: '/mock' });
+  const fakeHttpServices = fakeCoreStart.http as jest.Mocked<HttpSetup>;
+  fakeHttpServices.post.mockClear();
+  fakeHttpServices.get.mockClear();
+  fakeHttpServices.put.mockClear();
+  fakeHttpServices.delete.mockClear();
+
+  return fakeHttpServices;
+};
+
+export const render = async (
+  hook: any, // TODO: fix me
+  waitForHook: 'isSuccess' | 'isLoading' | 'isError' = 'isSuccess'
+) => {
+  const queryClient = new QueryClient();
+  const wrapper = ({ children }: { children: React.ReactNode }): JSX.Element => (
+    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
```

Comment:
Hey @paul-tavares , thanks for your feedback. Will answer your questions below:

> Why .invalidateQueries()'s in several of the hooks? It assumes a specific key has been used when the list hook was called, but the hook for list is also using options as part of the key. So I'm not sure how that is working

This will avoid us to manually refetch lists or get items after update/delete/create it. For example, when you add a new item, once the creation is done, you will have to update the list results, to include the new item. This is done by the invalidateQueries, you can invalidate whatever starts with an specific keys, so in this example from the updateArtifact:

```
 queryClient.invalidateQueries(['list', exceptionListApiClient]);
 queryClient.invalidateQueries(['get', exceptionListApiClient]);
```

It will invalidate (refetch) all the queries with keys starting with `list` or `get` and an specific `exceptionListApiClient` instance. See [this](https://react-query.tanstack.com/guides/query-invalidation#query-matching-with-invalidatequeries) for reference.

> Add return Types to all hooks (some were missing it)

Right, will do it! 👍 

>I noticed that you are passing the options object received on input directly to the queryKey array - might it be more efficient to destruct the object members and use that instead (which are normally primitive types of data)?

I'm ok to do that, my only concern is about the order when destructuring, items needs to be in specific order when you want to invalidate or clean cache using the query key, so destructuring the object items can make this a little more complex. Will try it

> Can we use the same exact component that is used in the real app? like that we ensure that the hooks are being tested within the same context as prod. code

This is used in our [app](https://github.com/elastic/kibana/blob/46eda470f5bd2136777559b008bd1530b43599ee/x-pack/plugins/security_solution/public/app/app.tsx). You mean using this app component instead of creating a new one for tests? I'm ok with that but I'm wondering if it will include too much logic.

Let me know what are your thoughts 🙂

Reply:
Thanks @dasansol92 .
I'll pinging you some of these topic offline. But,

> Re: distructing of options

I might not have been clear. What I mean was that the individual options should be used in the key rather than the object itself, like this:

```diff
  const { filter, page, perPage, policies } = options;

  return useQuery<FoundExceptionListItemSchema, ServerApiError>(
-   ['list', exceptionListApiClient, options],
+   ['list', exceptionListApiClient, filter, page, perPage, policies],
```


> Re: using the same `<QueryClientProvider...>` that is used in the app for tests

I did not realize that we added that context directly to the `app.tsx` file. I would suggest creating a separate component for that and then using it in both the `app.tsx` and in the test here. 

something like:

```typescript

export const queryClient = const queryClient = new QueryClient();

export const ReactQueryClientProvider = useMemo(({children}) => {
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
})
```

I'm also assuming that in the instance of `QueryClient` above is where we will be adding the default options, right? so when we do that for the app, they will also get picked up by the tests because we'll use the same Context Provider.

Reply:
> I'm also assuming that in the instance of QueryClient above is where we will be adding the default options, right? so when we do that for the app, they will also get picked up by the tests because we'll use the same Context Provider.

Yes, exactly. I'm still wondering if we should move this in a deeper level, maybe under management? Because this could be customised by each team...

Reply:
yeah, I remember that Esteban initially had it only under Management and I suggested it moving it up to the app because some of our APIs are also used in other areas of Security Solution

Being that we are the only ones using it now (I think), I would say leave it here. One can always just create one in their sub-app context and that would be used instead of the "global" one here in the `app.tsx